aggregator/handler/partitioned.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package handler
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/m3db/m3aggregator/aggregator"
+	"github.com/m3db/m3aggregator/partitioning"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errEmptyPartitions = errors.New("empty partitions")
+)
+
+// BackendServerPartition contains a backend server partition.
+type BackendServerPartition struct {
+	// Partition range.
+	Range partitioning.PartitionSet `validate:"nonzero"`
+
+	// Servers owning the partition.
+	Servers []string `validate:"nonzero"`
+}
+
+type partitionedHandlerMetrics struct {
+	partitionNotOwned tally.Counter
+}
+
+func newPartitionedHandlerMetrics(scope tally.Scope) partitionedHandlerMetrics {
+	return partitionedHandlerMetrics{
+		partitionNotOwned: scope.Counter("partition-not-owned"),
+	}
+}
+
+type handlerWithPartition struct {
+	Partitions partitioning.PartitionSet
+	Handler    aggregator.Handler
+}
+
+type partitionedHandler struct {
+	handlers []handlerWithPartition
+	metrics  partitionedHandlerMetrics
+}
+
+// NewPartitionedHandler creates a new partitioned handler.
+func NewPartitionedHandler(
+	partitions []BackendServerPartition,
+	opts Options,
+) (aggregator.Handler, error) {
+	numPartitions := len(partitions)
+	if numPartitions == 0 {
+		return nil, errEmptyPartitions
+	}
+	var (
+		// Divide overall queue size among all partitions.
+		partitionQueueSize = opts.QueueSize() / numPartitions
+		handlers           = make([]handlerWithPartition, 0, numPartitions)
+		instrumentOpts     = opts.InstrumentOptions()
+		scope              = instrumentOpts.MetricsScope()
+	)
+	for _, partition := range partitions {
+		handlerScope := scope.Tagged(map[string]string{
+			"min-partition": strconv.Itoa(partition.Range.Min()),
```

Comment:
If we start to put non-consecutive partitions in a partitionset then these 2 fields in this log would be not very helpful, how about we add a name/label to each partition and use the label for metrics/logging? 

For example if we createa a partition with range (e.g "0..50") then we use the range as the label, if we use dynamically generated m3cluster/placement, then we can perhaps use shardSetID as the label.

I imagine one day we might start supporting dynamic backends for indexers to save some deployments on the agg tier, so we should take that into consideration and try to stay closer to generic m3cluster placements.

Reply:
Sure can add a `Desc` or a `Name` to the `BackendServerPartition` struct and use that as the tag here.